server/src/main/java/org/elasticsearch/search/lookup/LeafDocLookup.java

```diff
@@ -84,22 +144,31 @@ public DocValuesScriptFieldFactory run() {
             throw ExceptionsHelper.convertToElastic(ioe);
         }
 
-        return factory;
-    }
-
-    public Field<?> getScriptField(String fieldName) {
-        return getScriptFieldFactory(fieldName, MappedFieldType.FielddataOperation.SCRIPT).toScriptField();
+        return factory.toScriptField();
     }
 
     @Override
     public ScriptDocValues<?> get(Object key) {
-        return getScriptFieldFactory(key.toString(), MappedFieldType.FielddataOperation.SEARCH).toScriptDocValues();
+        String fieldName = key.toString();
+        DocValuesScriptFieldFactory factory = docFactoryCache.get(fieldName);
+
+        if (factory == null) {
+            factory = getScriptFieldFactory(key.toString(), false);
+        }
+
+        try {
+            factory.setNextDocId(docId);
+        } catch (IOException ioe) {
+            throw ExceptionsHelper.convertToElastic(ioe);
+        }
+
+        return factory.toScriptDocValues();
     }
 
     @Override
     public boolean containsKey(Object key) {
         String fieldName = key.toString();
-        return localCacheScriptFieldData.get(fieldName) != null || fieldTypeLookup.apply(fieldName) != null;
+        return fieldTypeLookup.apply(fieldName) != null;
```

Comment:
Why can this method skip checking the caches?

Reply:
I'm actually not sure why the caches were involved at all. fieldTypeLookup contains all the available mappings and since both end up being hash lookups it seems like extra overhead in general.

Reply:
I actually just double-checked and it appears there's actually some work involved in the function to get the field type, so I'll go ahead and add both caches with the doc-style access first as it's more likely useful in that situation. Good catch.